### ISSUE TITLE
Symbols remain defined after failure and backtrack

### ISSUE THREAD

**Issue by ChrisHixon** [yhirose/cpp-peglib#231]
This is a contrived example to test how symbols behave when set during a rule match that fails, causing backtrack. I would expect the failure and backtrack to undo the symbol being set.

I dealt with similar issues when I added back-references to chpeg. I ended up creating an 'undo action' system that allows actions to be run upon failure/backtrack to undo what has been done. Perhaps this is something to consider for cpp-peglib. My initial back-reference implementation did something like what cpp-peglib appears to do: push a back reference stack and copy results back upon success. This seems inefficient and has to occur before/after every operation that can possibly set a reference.

Grammar:
```
S            <- (DeclBT / Decl / Ref)*
DeclBT       <- 'decl' symbol 'backtrack' # match fails, so symbol should not be set
Decl         <- 'decl' symbol
Ref          <- 'ref' is_symbol
Name         <- < [a-zA-Z]+ >
%whitespace  <- [ \t\r\n]*

# 'var_table' is a table name.
symbol       <- Name { declare_symbol var_table } # Declare symbol instruction
is_symbol    <- Name { check_symbol var_table }   # Check symbol instruction
```
Input:
```
decl foo
ref foo
```
Result:
```
1:6 'foo' already exists.
```
Expected: no failure

**Comment by yhirose** [comment]
@ChrisHixon, thanks for reporting the bug! I'll look into it.

**Comment by mingodad** [comment]
Somehow relate this project has a good description of desired features in a parser https://github.com/Lercher/CocoR that it somehow implements (worth short reading).

**Comment by yhirose** [comment]
I fixed the reported problem as well as `S  <- DeclBT* Decl Ref`. However, there are the following remaining bugs:

 *  This feature doesn't work with the packrat mode because the packrat table doesn't record

**Comment by yhirose** [comment]
Closed by #233
